File: bench/benches/closure.exs
Code.require_file("../lib/bench.exs", __DIR__)
# closures escape into a list, then applied through it (non-inlinable). checksum = sum 3i.
twice = fn f -> fn x -> f.(f.(x)) end end
adder = fn i -> fn x -> x + i end end
Bench.run("closure", fn ->
  fns = Enum.map(0..99999, fn i -> twice.(adder.(i)) end)
  fns |> Enum.with_index() |> Enum.reduce(0, fn {f, i}, s -> s + f.(i) end)
end)

File: bench/benches/closure.rs
include!("../lib/bench.rs");
fn adder(i: i64) -> impl Fn(i64) -> i64 {
    move |x| x + i
}
fn twice(f: impl Fn(i64) -> i64) -> impl Fn(i64) -> i64 {
    move |x| f(f(x))
}
const N: i64 = 100000;
const M: i64 = 1000000007;
fn main() {
    bench("closure", || {
        let mut acc: i64 = 0;
        for i in 0..N {
            acc = (acc * 31 + twice(adder(i))(i)) % M;
        }
        acc
    });
}
